src/agent/thread_helper.cpp

```diff
@@ -0,0 +1,323 @@
+#include "agent/thread_helper.hpp"
+
+#include <assert.h>
+#include <string.h>
+
+#include <string>
+
+#include "openthread/border_router.h"
+#include "openthread/channel_manager.h"
+#include "openthread/jam_detection.h"
+#include "openthread/joiner.h"
+#include "openthread/platform/radio.h"
+#include "openthread/thread_ftd.h"
+
+#include "agent/ncp_openthread.hpp"
+#include "common/code_utils.hpp"
+
+namespace otbr {
+namespace agent {
+
+ThreadHelper::ThreadHelper(otInstance *aInstance, ot::BorderRouter::Ncp::ControllerOpenThread *aNcp)
+    : mInstance(aInstance)
+    , mNcp(aNcp)
+    , mScanHandler(nullptr)
+{
+}
+
+otError ThreadHelper::Init(void)
+{
+    srand(static_cast<unsigned int>(time(nullptr)));
+
+    return otSetStateChangedCallback(mInstance, sStateChangedCallback, this);
+}
+
+void ThreadHelper::sStateChangedCallback(otChangedFlags aFlags, void *aContext)
+{
+    ThreadHelper *network = reinterpret_cast<ThreadHelper *>(aContext);
+
+    network->StateChangedCallback(aFlags);
+}
+
+void ThreadHelper::StateChangedCallback(otChangedFlags aFlags)
+{
+    if (aFlags & OT_CHANGED_THREAD_ROLE)
+    {
+        otDeviceRole role = otThreadGetDeviceRole(mInstance);
+
+        for (auto &&f : mDeviceRoleHandlers)
+        {
+            f(role);
+        }
+
+        if (role == OT_DEVICE_ROLE_DISABLED && mLeaveHandler != nullptr)
+        {
+            mLeaveHandler(OT_ERROR_NONE);
+            mLeaveHandler = nullptr;
+        }
+
+        if (role != OT_DEVICE_ROLE_DISABLED && role != OT_DEVICE_ROLE_DETACHED)
+        {
+            if (mAttachHandler != nullptr)
+            {
+                mAttachHandler(OT_ERROR_NONE);
+                mAttachHandler = nullptr;
+            }
+            else if (mJoinerHandler != nullptr)
+            {
+                mJoinerHandler(OT_ERROR_NONE);
+                mJoinerHandler = nullptr;
+            }
+        }
+    }
+    (void)aFlags;
+}
+
+void ThreadHelper::AddDeviceRoleHandler(const DeviceRoleHandler &aHandler)
+{
+    mDeviceRoleHandlers.insert(mDeviceRoleHandlers.end(), aHandler);
+}
+
+otError ThreadHelper::Scan(const ScanHandler &aHandler)
+{
+    otError err;
+
+    mScanHandler = aHandler;
+    mScanResults.clear();
+
+    err = otLinkActiveScan(mInstance, /*scanChannels =*/0, /*scanDuration=*/0, &ThreadHelper::sActiveScanHandler, this);
+
+    if (err != OT_ERROR_NONE)
+    {
+        mScanHandler = nullptr;
+    }
+
+    return err;
+}
+
+static void RandomFill(void *aBuf, size_t size)
+{
+    uint8_t *buf = reinterpret_cast<uint8_t *>(aBuf);
+    for (size_t i = 0; i < size; i++)
+    {
+        buf[i] = rand() & 0xff;
```

Comment:
The quality of numbers generated by `rand` is not guaranteed. If possible, use `random_device` or The Abseil Random Library instead.